11.bulk_Morphology_Elastic_Network_LOCO/2.test_models/scripts/1.test_regression_multi_output.py

```diff
@@ -0,0 +1,331 @@
+        "log10_explained_variance",
+    ]
+)
+
+
+# In[13]:
+
+
+data_x = data_split_x
+data_y = data_split_y
+if shuffle == "shuffled_baseline":
+    model = joblib.load(
+        f"../../1.train_models/{model_path}/{cytokine}_{feature_combinations_key}_shuffled_baseline__all_nomic.joblib"
+    )
+elif shuffle == "final":
+    model = joblib.load(
+        f"../../1.train_models/{model_path}/{cytokine}_{feature_combinations_key}_final__all_nomic.joblib"
+    )
+else:
+    print("Error")
+
+# get the cytokine column of choice
+y_selected = data_y[cytokine]
+
+if shuffle == "shuffled_baseline":
+    for column in data_x:
+        np.random.shuffle(data_x[column].values)
+
+# get predictions
+predictions = model.predict(data_x)
+
+
+# In[14]:
+
+
+explained_variance = explained_variance_score(y_selected, predictions)
+output_metric_scores["explained_variance"] = explained_variance
+neg_mean_absolute_error = -mean_squared_error(y_selected, predictions)
+output_metric_scores["neg_mean_absolute_error"] = neg_mean_absolute_error
+neg_mean_squared_error = -mean_squared_error(y_selected, predictions)
+output_metric_scores["neg_mean_squared_error"] = neg_mean_squared_error
+r2 = r2_score(y_selected, predictions)
+output_metric_scores["treatment"] = metadata[
+    "oneb_Metadata_Treatment_Dose_Inhibitor_Dose"
+]
+output_metric_scores["well"] = metadata["Metadata_Well"].values
+df = pd.DataFrame.from_dict(output_metric_scores)
+df["r2"] = r2
+df["cytokine"] = cytokine
+df["data_split"] = data_split
+df["shuffle"] = shuffle
+df["predicted_value"] = predictions
+df["actual_value"] = y_selected
+df["log10_neg_mean_absolute_error"] = -np.log10(-df["neg_mean_absolute_error"])
+df["log10_neg_mean_squared_error"] = -np.log10(-df["neg_mean_squared_error"])
+df["log10_explained_variance"] = -np.log10(df["explained_variance"])
+
+# replace "[NSU]" with """
+df["cytokine"] = df["cytokine"].replace("[ \[\]NSU]", "", regex=True)
+df["cytokine"] = df["cytokine"].replace(" ", "_", regex=True)
+
+# concat the dataframes
+results_df = pd.concat([results_df, df], axis=0)
+results_df["channel_feature_combinations_key"] = feature_combinations_key
+
+
+# In[15]:
+
+
+results_df.head()
+
+
+# In[16]:
+
+
+var_df = results_df.drop(
+    columns=[
+        "explained_variance",
+        "neg_mean_absolute_error",
+        "neg_mean_squared_error",
+        "well",
+        "treatment",
+        "r2",
+        "log10_neg_mean_absolute_error",
+        "log10_neg_mean_squared_error",
+        "log10_explained_variance",
+    ]
+)
+# calculate the variance of the actual and predicted values per cytokine
+var_df = var_df.groupby(
+    ["cytokine", "data_split", "shuffle", "channel_feature_combinations_key"]
+).var()
+var_df = pd.merge(
+    var_df,
+    results_df.groupby(
+        ["cytokine", "data_split", "shuffle", "channel_feature_combinations_key"]
+    ).r2.unique(),
+    left_index=True,
+    right_index=True,
+)
+var_df.reset_index(inplace=True)
+var_df.head()
```

Comment:
Ah, now I see what you're doing. But the question remains whether you should add comments to help other readers to understand.